src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -111,5 +119,106 @@ class EntityClient (requestContext: RequestContext) extends Actor {
+          requestContext.complete( HttpResponse(BadRequest, "TSV is missing required attributes: " + (requiredAttributes.keySet -- headers).mkString(", ")) )
+        } else {
+          op(requiredAttributes)
+        }
+      }
+    }
+  }
+
+  def addToCollectionMembersAttribute( entityType: String, entityName: String, newMembers: Seq[(String, String)], membersAttributeName: String ) = {
+    //STUB: return a partial batch Rawls call for this:
+    //(entityName, entityType).attrs[membersAttributeName].extend(newMembers)
+  }
+
+  def setAttributesOnEntity(entityType: String, entityName: String, attributesMap: Map[String, String], referenceTypes: Map[String, String] ) = {
+    //for all attributes in attributesMap
+    // if the key is present in referenceTypes, the value is an entity name and referenceTypes(key) contains that entity's type
+    //MAYBE-DO: Parse out types provided in the TSV header? Or allow the model to specify types for optionally specified attributes?
+
+    //STUB: return a batch of AddUpdateOperations
+  }
+
+  def batchCallToRawls( calls: Any ) = {
+    //STUB: stitch the batch calls together and fire off the Rawls request.
+    //requestContext.complete(...)
+  }
+
+  def addToCollectionTypeFromTSV( workspaceNamespace: String, workspaceName: String, entityType: String, memberType: String, membersAttributeName: String, tsv: TSVLoadFile ) = {
+    //group together updates to distinct entities, construct the list of entities to add references to, and poke it into whatever the "members" field is called.
+    val tsvFieldName = memberType + "_id"
+    tsv.tsvData.groupBy( _(tsv.firstColumnHeader) ).map({
+      case (entityName, rows) => addToCollectionMembersAttribute( entityType, entityName, rows.map(row => (memberType, row(tsvFieldName))), membersAttributeName )
+    })
+  }
+
+  //TSVs can be of two forms: they can either populate the "members" list of a collection entity,
+  //or they can populate the attributes on an(y kind of) entity.
+  def upsertEntitiesFromTSV(workspaceNamespace: String, workspaceName: String, tsv: TSVLoadFile): Unit = {
+    //the first column header of the tsv defines the entity type we're trying to import or update.
+    //This is magically verified in the withFoo functions below.
+    val entityType = tsv.firstColumnHeader.stripSuffix("_id")
+
+    withMemberCollectionType(entityType) { memberTypeOpt =>
+      //This magical list of conditions determines whether the TSV is populating the "members" attribute of a collection type entity.
+      if( memberTypeOpt.isDefined
+          && tsv.headers.length == 2
+          && tsv.headers == Seq( tsv.firstColumnHeader, memberTypeOpt.get + "_id" ) ) {
+
+        withPlural(memberTypeOpt.get) { memberPlural =>
+          val rawlsCalls = addToCollectionTypeFromTSV(workspaceNamespace, workspaceName, entityType, memberTypeOpt.get, memberPlural, tsv)
+          batchCallToRawls(rawlsCalls)
+        }
+      } else {
```

Comment:
should this be an else, or are there cases where the above might not be true due to user error and we will try to proceed (should it instead be an else if?)